src/main/java/com/android/volley/Request.java

```diff
@@ -320,17 +344,81 @@ public String getCacheKey() {
     }
 
     /**
-     * Mark this request as canceled.  No callback will be delivered.
+     * Mark this request as canceled.
+     *
+     * <p>No callback will be delivered as long as {@link ExecutorDelivery} is used as the
+     * {@link ResponseDelivery} and as long as this method is not overridden. There are no
+     * guarantees otherwise.
+     *
+     * <p>Subclasses should generally override {@link #onCanceled()} instead to perform an action
+     * when the request is canceled (such as clearing response listeners to avoid leaks).
      */
     public void cancel() {
-        mCanceled = true;
+        synchronized (mLock) {
+            mCanceled = true;
+            onCanceledInternal();
+        }
     }
 
     /**
      * Returns true if this request has been canceled.
      */
     public boolean isCanceled() {
-        return mCanceled;
+        synchronized (mLock) {
+            return mCanceled;
+        }
+    }
+
+    private void onCanceledInternal() {
+        mErrorListener = null;
+        if (mCancelListener != null) {
+            mCancelListener.onRequestCanceled();
+            mCancelListener = null;
+        }
+        onCanceled();
+    }
+
+    /**
+     * Called when a request is canceled.
+     *
+     * <p>The default implementation does nothing. Subclasses may override this to perform custom
+     * actions, such as clearing response listeners to avoid leaks.
+     */
+    protected void onCanceled() {
+    }
+
+    /** Deliver the given response as long as the request isn't canceled. */
+    void deliverResponse(Response<T> response, Runnable runnable) {
+        // To uphold the contract of cancel(), we hold the same lock here around both the check of
+        // isCanceled() as well as the actual response delivery. This ensures that after a call
+        // to cancel() returns, we will never deliver the response.
+        synchronized (mLock) {
```

Comment:
The change from open calls (without lock held) to closed calls makes me nervous about deadlock.